// patronum/debug/index.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.debug = debug;

var _effector = require("effector");

function debug(...units) {
  for (const unit of units) {
    const type = getType(unit);

    if (_effector.is.store(unit, {
      sid: "8mdztw"
    }) || _effector.is.effect(unit, {
      sid: "-gueih2"
    }) || _effector.is.event(unit, {
      sid: "-gueh0j"
    })) {
      log(unit, type);
    }

    if (_effector.is.effect(unit, {
      sid: "-jlpzbv"
    })) {
      logEffect(unit);
    }

    if (_effector.is.domain(unit, {
      sid: "-jjisyf"
    })) {
      unit.onCreateEvent(event => log(event, 'event'));
      unit.onCreateStore(store => log(store, 'store'));
      unit.onCreateEffect(logEffect);
    }
  }
}

function getType(unit) {
  if (_effector.is.store(unit, {
    sid: "-j30z7b"
  })) {
    return 'store';
  }

  if (_effector.is.effect(unit, {
    sid: "-j1dlf8"
  })) {
    return 'effect';
  }

  if (_effector.is.event(unit, {
    sid: "-io6j6k"
  })) {
    return 'event';
  }

  if (_effector.is.domain(unit, {
    sid: "-imj5eh"
  })) {
    return 'domain';
  }

  if (_effector.is.unit(unit)) {
    return 'unit';
  }

  return 'unknown';
}

function log(unit, type, prefix = '') {
  const name = prefix + getName(unit);
  unit.watch(payload => {
    console.info(`[${type}] ${name}`, payload);
  });
}

function logEffect(unit) {
  log(unit.done, 'effect', getName(unit) + '.');
  log(unit.fail, 'effect', getName(unit) + '.');
}

function getName(unit) {
  if (unit.compositeName && unit.compositeName.fullName) {
    return unit.compositeName.fullName;
  }

  if (unit.shortName) {
    return unit.shortName;
  }

  if (unit.name) {
    return unit.name;
  }

  return '';
}
